In [0]:
pip install pomegranate

     |████████████████████████████████| 3.3MB 3.9MB/s 
  Created wheel for pomegranate: filename=pomegranate-0.12.2-cp36-cp36m-linux_x86_64.whl size=10965170 sha256=61362d679245199620b998c2c1a0aac33423c54318c745909d6499d9b0873ac5
  Stored in directory: /root/.cache/pip/wheels/51/14/41/7285a60a08951abde03ce60f47a8644dd96d8e22b373bbbcc8
Successfully built pomegranate


In [0]:
pip install ipython-autotime

  Created wheel for ipython-autotime: filename=ipython_autotime-0.1-cp36-none-any.whl size=1832 sha256=8f891e641d93b5ef7ead50744d9a528834e956e2b52de89966ae094f07ab69f3
  Stored in directory: /root/.cache/pip/wheels/d2/df/81/2db1e54bc91002cec40334629bc39cfa86dff540b304ebcd6e
Successfully built ipython-autotime


In [0]:
%load_ext autotime

In [0]:
import time
import numpy as np
import pandas as pd
from pomegranate import *
import matplotlib.pyplot as plt
import random

time: 3.99 s


In [0]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
time: 1.71 ms


In [0]:
dataset = pd.read_csv("/content/drive/My Drive/DAMIN/diabetes.csv")
dataset

Pregnancies  Glucose  ...  Age  Outcome
0              6      148  ...   50        1
1              1       85  ...   31        0
2              8      183  ...   32        1
3              1       89  ...   21        0
4              0      137  ...   33        1
..           ...      ...  ...  ...      ...
763           10      101  ...   63        0
764            2      122  ...   27        0
765            5      121  ...   30        0
766            1      126  ...   47        1
767            1       93  ...   23        0

[768 rows x 9 columns]

time: 45.2 ms


In [0]:
#Hanya gunakan non categorical values
df = dataset.drop(['Outcome'], axis=1)
df

Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
0              6      148             72  ...  33.6                     0.627   50
1              1       85             66  ...  26.6                     0.351   31
2              8      183             64  ...  23.3                     0.672   32
3              1       89             66  ...  28.1                     0.167   21
4              0      137             40  ...  43.1                     2.288   33
..           ...      ...            ...  ...   ...                       ...  ...
763           10      101             76  ...  32.9                     0.171   63
764            2      122             70  ...  36.8                     0.340   27
765            5      121             72  ...  26.2                     0.245   30
766            1      126             60  ...  30.1                     0.349   47
767            1       93             70  ...  30.4                     0.315   23

[768 rows x 8 columns]

time: 33 ms


In [0]:
X = df.values
df

Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
0              6      148             72  ...  33.6                     0.627   50
1              1       85             66  ...  26.6                     0.351   31
2              8      183             64  ...  23.3                     0.672   32
3              1       89             66  ...  28.1                     0.167   21
4              0      137             40  ...  43.1                     2.288   33
..           ...      ...            ...  ...   ...                       ...  ...
763           10      101             76  ...  32.9                     0.171   63
764            2      122             70  ...  36.8                     0.340   27
765            5      121             72  ...  26.2                     0.245   30
766            1      126             60  ...  30.1                     0.349   47
767            1       93             70  ...  30.4                     0.315   23

[768 rows x 8 columns]

time: 31.4 ms


In [0]:
df.isna().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
dtype: int64

time: 6 ms


In [0]:
#drop jika ada NaN value
if (df.isna().values.any()):
  df.dropna()

time: 2.72 ms


In [0]:
X.shape

(768, 8)

time: 2.81 ms


#KNN

In [0]:
#membuat NaN values ke dataset
random.seed(0)
np.random.seed(0)
np.set_printoptions(suppress=True)

idxs = numpy.random.choice(768, replace=False, size=15)
j =  np.random.choice(range(3,8),15)
X_nan = X.copy()
print(idxs)
print(j)

[661 122 113  14 529 103 338 588 395 204  31 546 278 593 737]
[6 7 7 3 3 5 7 6 6 3 3 4 7 7 3]
time: 6 ms


In [0]:
X_nan[idxs, j] = np.nan

time: 1.37 ms


In [0]:
from sklearn.impute import KNNImputer

imputer = KNNImputer(n_neighbors=5)
df_filled = imputer.fit_transform(X_nan)
df_filled

array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],
       [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],
       [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],
       [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],
       [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]])

time: 239 ms


In [0]:
#check error
import sklearn.metrics as metrics

mae = metrics.mean_absolute_error(X, df_filled)
mse = metrics.mean_squared_error(X, df_filled)
rmse = np.sqrt(mse) # or mse**(0.5)

print("Results of sklearn.metrics:")
print("MAE:",mae)
print("MSE:", mse)
print("RMSE:", rmse)

Results of sklearn.metrics:
MAE: 0.021501009114583327
MSE: 0.668684988938802
RMSE: 0.817731611801086
time: 8.54 ms


#Bayesian

In [0]:
#membuat NaN values ke dataset
random.seed(0)
np.random.seed(0)
np.set_printoptions(suppress=True)

idxs = numpy.random.choice(768, replace=False, size=15)
j =  np.random.choice(range(3,8),15)
X_nan = X.copy()
print(idxs)
print(j)

X_nan[idxs, j] = numpy.nan

[661 122 113  14 529 103 338 588 395 204  31 546 278 593 737]
[6 7 7 3 3 5 7 6 6 3 3 4 7 7 3]
time: 6.47 ms


In [0]:
X_nan.shape

(768, 8)

time: 2.53 ms


In [0]:
model1 = BayesianNetwork.from_samples(X_nan, max_parents=2)

time: 535 ms


In [0]:
model1.predict_proba({})

array([{
    "class" :"Distribution",
    "dtype" :"numpy.float64",
    "name" :"DiscreteDistribution",
    "parameters" :[
        {
            "6.0" :0.06510416666666666,
            "1.0" :0.17578124999999958,
            "8.0" :0.049479166666666706,
            "0.0" :0.1445312499999997,
            "5.0" :0.07421874999999996,
            "3.0" :0.09765624999999986,
            "10.0" :0.031250000000000104,
            "2.0" :0.13411458333333307,
            "4.0" :0.08854166666666656,
            "7.0" :0.05859375000000001,
            "9.0" :0.036458333333333426,
            "11.0" :0.014322916666666836,
            "13.0" :0.013020833333333507,
            "15.0" :0.0013020833333335505,
            "17.0" :0.0013020833333335505,
            "12.0" :0.011718750000000179,
            "14.0" :0.002604166666666879
        }
    ],
    "frozen" :false
},
       {
    "class" :"Distribution",
    "dtype" :"numpy.float64",
    "name" :"DiscreteDistribution",
    "parameters" :[
      

time: 37 ms


In [0]:
y_pred1 = model1.predict(X_nan)

time: 11.2 s


In [0]:
#check error
import sklearn.metrics as metrics

mae = metrics.mean_absolute_error(X, np.array(y_pred1))
mse = metrics.mean_squared_error(X, np.array(y_pred1))
rmse = np.sqrt(mse) # or mse**(0.5)

print("Results of sklearn.metrics:")
print("MAE:",mae)
print("MSE:", mse)
print("RMSE:", rmse)

Results of sklearn.metrics:
MAE: 0.0616513671875
MSE: 7.724134328776041
RMSE: 2.7792326870515973
time: 6.78 ms


#Deterministic Regression

In [0]:
df2 = dataset.drop(['Outcome'], axis=1)
X2 = df2.values
df2

Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
0              6      148             72  ...  33.6                     0.627   50
1              1       85             66  ...  26.6                     0.351   31
2              8      183             64  ...  23.3                     0.672   32
3              1       89             66  ...  28.1                     0.167   21
4              0      137             40  ...  43.1                     2.288   33
..           ...      ...            ...  ...   ...                       ...  ...
763           10      101             76  ...  32.9                     0.171   63
764            2      122             70  ...  36.8                     0.340   27
765            5      121             72  ...  26.2                     0.245   30
766            1      126             60  ...  30.1                     0.349   47
767            1       93             70  ...  30.4                     0.315   23

[768 rows x 8 columns]

time: 34.4 ms


In [0]:
#membuat NaN values ke dataset
random.seed(0)
np.random.seed(0)
np.set_printoptions(suppress=True)

idxs = numpy.random.choice(768, replace=False, size=15)
j =  np.random.choice(range(3,8),15)
X_nan = X.copy()
print(idxs)
print(j)
X_nan[idxs, j] = numpy.nan
df_nan = pd.DataFrame(X_nan, columns=['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age'])
df_nan

[661 122 113  14 529 103 338 588 395 204  31 546 278 593 737]
[6 7 7 3 3 5 7 6 6 3 3 4 7 7 3]


Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction   Age
0            6.0    148.0           72.0  ...  33.6                     0.627  50.0
1            1.0     85.0           66.0  ...  26.6                     0.351  31.0
2            8.0    183.0           64.0  ...  23.3                     0.672  32.0
3            1.0     89.0           66.0  ...  28.1                     0.167  21.0
4            0.0    137.0           40.0  ...  43.1                     2.288  33.0
..           ...      ...            ...  ...   ...                       ...   ...
763         10.0    101.0           76.0  ...  32.9                     0.171  63.0
764          2.0    122.0           70.0  ...  36.8                     0.340  27.0
765          5.0    121.0           72.0  ...  26.2                     0.245  30.0
766          1.0    126.0           60.0  ...  30.1                     0.349  47.0
767          1.0     93.0           70.0  ...  30.4                     0.315  23.0

[768 rows x 8 columns]

time: 42 ms


In [0]:
df_nan.isna().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               5
Insulin                     1
BMI                         1
DiabetesPedigreeFunction    3
Age                         5
dtype: int64

time: 5.22 ms


In [0]:
missing_columns = ['SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']
deter_data = pd.DataFrame(columns = ["Det" + name for name in missing_columns])
from sklearn.linear_model import LinearRegression
for feature in missing_columns:
  # print(feature)
  deter_data["Det" + feature] = df_nan[feature]
  parameters = list(set(df_nan.columns) - set(missing_columns) - {feature})

  # #Create a Linear Regression model to estimate the missing data
  model = LinearRegression()
  model.fit(X = df2[parameters], y = df2[feature])

  #observe that I preserve the index of the missing data from the original dataframe
  deter_data.loc[df_nan[feature].isna(), "Det" + feature] = model.predict(df_nan[parameters])[df_nan[feature].isna()]

time: 52.8 ms


In [0]:
deter_data

DetSkinThickness  DetInsulin  DetBMI  DetDiabetesPedigreeFunction  DetAge
0                35.0         0.0    33.6                        0.627    50.0
1                29.0         0.0    26.6                        0.351    31.0
2                 0.0         0.0    23.3                        0.672    32.0
3                23.0        94.0    28.1                        0.167    21.0
4                35.0       168.0    43.1                        2.288    33.0
..                ...         ...     ...                          ...     ...
763              48.0       180.0    32.9                        0.171    63.0
764              27.0         0.0    36.8                        0.340    27.0
765              23.0       112.0    26.2                        0.245    30.0
766               0.0         0.0    30.1                        0.349    47.0
767              31.0         0.0    30.4                        0.315    23.0

[768 rows x 5 columns]

time: 26.1 ms


In [0]:
non_missing_columns = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']
data_non_nan = pd.DataFrame(columns = ["Det" + name for name in non_missing_columns])

for feature in non_missing_columns:
  data_non_nan["Det" + feature] = df2[feature]

for feature in missing_columns:
  data_non_nan["Det" + feature] = deter_data["Det" + feature]
data_non_nan

DetPregnancies  DetGlucose  ...  DetDiabetesPedigreeFunction  DetAge
0                 6         148  ...                        0.627    50.0
1                 1          85  ...                        0.351    31.0
2                 8         183  ...                        0.672    32.0
3                 1          89  ...                        0.167    21.0
4                 0         137  ...                        2.288    33.0
..              ...         ...  ...                          ...     ...
763              10         101  ...                        0.171    63.0
764               2         122  ...                        0.340    27.0
765               5         121  ...                        0.245    30.0
766               1         126  ...                        0.349    47.0
767               1          93  ...                        0.315    23.0

[768 rows x 8 columns]

time: 46.1 ms


In [0]:
#check error
import sklearn.metrics as metrics

mae = metrics.mean_absolute_error(X, data_non_nan.values)
mse = metrics.mean_squared_error(X, data_non_nan.values)
rmse = np.sqrt(mse) # or mse**(0.5)

print("Results of sklearn.metrics:")
print("MAE:",mae)
print("MSE:", mse)
print("RMSE:", rmse)

Results of sklearn.metrics:
MAE: 0.02554667542880195
MSE: 0.6424961416906976
RMSE: 0.8015585703432392
time: 7.43 ms


#Summary

##Diabetes

###KNN


####Results of sklearn.metrics:
####MAE: 0.021501009114583327
####MSE: 0.668684988938802
####RMSE: 0.817731611801086

###Bayesian Network

####Results of sklearn.metrics:
####MAE: 0.0616513671875
####MSE: 7.724134328776041
####RMSE: 2.7792326870515973

###Deterministic Regression

####Results of sklearn.metrics:
####MAE: 0.02554667542880197
####MSE: 0.6424961416906995
####RMSE: 0.8015585703432404